// @videojs/react/dist/dev/ui/tooltip/tooltip-popup.js
'use client';

import { useComposedRefs } from "../../utils/use-composed-refs.js";
import { renderElement } from "../../utils/use-render.js";
import { useTooltipContext } from "./context.js";
import { getAnchorPositionStyle, getPopupPositionRect, resolveOffsets } from "@videojs/core/dom";
import { forwardRef, useCallback, useLayoutEffect, useMemo, useRef, useState } from "react";
import { TooltipCSSVars } from "@videojs/core";
import { supportsAnchorPositioning } from "@videojs/utils/dom";

//#region src/ui/tooltip/tooltip-popup.tsx
const POPUP_RESET = {
	position: "fixed",
	inset: "auto",
	margin: 0
};
/** Container for the tooltip content. Positioned relative to the trigger using CSS anchor positioning with a JavaScript fallback. */
const TooltipPopup = forwardRef(function TooltipPopup({ render, className, style, ...elementProps }, forwardedRef) {
	const { core, tooltip, state, stateAttrMap, anchorName, popupId } = useTooltipContext();
	const internalRef = useRef(null);
	const composedRef = useComposedRefs(forwardedRef, useCallback((el) => {
		tooltip.setPopupElement(el);
		if (el && supportsAnchorPositioning()) el.style.setProperty("position-anchor", `--${anchorName}`);
	}, [tooltip, anchorName]), internalRef);
	const posOpts = useMemo(() => ({
		side: state.side,
		align: state.align
	}), [state.side, state.align]);
	const anchorStyle = useMemo(() => {
		if (!supportsAnchorPositioning()) return null;
		const { positionAnchor: _, ...rest } = getAnchorPositionStyle(anchorName, posOpts, void 0, void 0, void 0, void 0, TooltipCSSVars);
		return rest;
	}, [anchorName, posOpts]);
	const [manualStyle, setManualStyle] = useState(null);
	useLayoutEffect(() => {
		if (supportsAnchorPositioning()) return;
		if (!state.open) {
			setManualStyle(null);
			return;
		}
		function measure() {
			const triggerEl = tooltip.triggerElement;
			const popupEl = internalRef.current;
			if (!triggerEl || !popupEl) return;
			setManualStyle(getAnchorPositionStyle(anchorName, posOpts, triggerEl.getBoundingClientRect(), getPopupPositionRect(popupEl), document.documentElement.getBoundingClientRect(), resolveOffsets(popupEl, TooltipCSSVars), TooltipCSSVars));
		}
		measure();
		const triggerEl = tooltip.triggerElement;
		const popupEl = internalRef.current;
		let rafId = 0;
		function reposition() {
			cancelAnimationFrame(rafId);
			rafId = requestAnimationFrame(measure);
		}
		reposition();
		const resizeObserver = typeof ResizeObserver === "function" ? new ResizeObserver(() => {
			reposition();
		}) : null;
		if (triggerEl && resizeObserver) resizeObserver.observe(triggerEl);
		if (popupEl && resizeObserver) resizeObserver.observe(popupEl);
		window.addEventListener("scroll", reposition, {
			capture: true,
			passive: true
		});
		window.addEventListener("resize", reposition);
		return () => {
			cancelAnimationFrame(rafId);
			resizeObserver?.disconnect();
			window.removeEventListener("scroll", reposition, true);
			window.removeEventListener("resize", reposition);
		};
	}, [
		state.open,
		anchorName,
		posOpts,
		tooltip
	]);
	const positioningStyle = anchorStyle ?? manualStyle ?? POPUP_RESET;
	if (!state.open) return null;
	const { onFocusOut, ...restPopupProps } = tooltip.popupProps;
	return renderElement("div", {
		render,
		className,
		style
	}, {
		state,
		stateAttrMap,
		ref: composedRef,
		props: [
			{
				id: popupId,
				style: positioningStyle,
				...core.getPopupAttrs(state)
			},
			{
				...restPopupProps,
				onBlur: onFocusOut
			},
			elementProps
		]
	});
});

//#endregion
export { TooltipPopup };
