// @videojs/react/dist/dev/presets/video/skin.js
import { Container, usePlayer } from "../../player/context.js";
import { isRenderProp } from "../../utils/use-render.js";
import { BufferingIndicator } from "../../ui/buffering-indicator/buffering-indicator.js";
import { CaptionsButton } from "../../ui/captions-button/captions-button.js";
import { ControlsRoot } from "../../ui/controls/controls-root.js";
import { FullscreenButton } from "../../ui/fullscreen-button/fullscreen-button.js";
import { MuteButton } from "../../ui/mute-button/mute-button.js";
import { PiPButton } from "../../ui/pip-button/pip-button.js";
import { PlayButton } from "../../ui/play-button/play-button.js";
import { PlaybackRateButton } from "../../ui/playback-rate-button/playback-rate-button.js";
import { PopoverPopup } from "../../ui/popover/popover-popup.js";
import { PopoverRoot } from "../../ui/popover/popover-root.js";
import { PopoverTrigger } from "../../ui/popover/popover-trigger.js";
import { Poster } from "../../ui/poster/poster.js";
import { SeekButton } from "../../ui/seek-button/seek-button.js";
import { SliderBuffer } from "../../ui/slider/slider-buffer.js";
import { SliderFill } from "../../ui/slider/slider-fill.js";
import { SliderThumb } from "../../ui/slider/slider-thumb.js";
import { SliderThumbnail } from "../../ui/slider/slider-thumbnail.js";
import { SliderTrack } from "../../ui/slider/slider-track.js";
import { SliderValue } from "../../ui/slider/slider-value.js";
import { Group } from "../../ui/time/time-group.js";
import { Value } from "../../ui/time/time-value.js";
import { TimeSliderRoot } from "../../ui/time-slider/time-slider-root.js";
import { TooltipPopup } from "../../ui/tooltip/tooltip-popup.js";
import { TooltipProvider } from "../../ui/tooltip/tooltip-provider.js";
import { TooltipRoot } from "../../ui/tooltip/tooltip-root.js";
import { TooltipTrigger } from "../../ui/tooltip/tooltip-trigger.js";
import { VolumeSliderRoot } from "../../ui/volume-slider/volume-slider-root.js";
import CaptionsOffIcon from "../../icons/dist/react/default/captions-off.js";
import CaptionsOnIcon from "../../icons/dist/react/default/captions-on.js";
import FullscreenEnterIcon from "../../icons/dist/react/default/fullscreen-enter.js";
import FullscreenExitIcon from "../../icons/dist/react/default/fullscreen-exit.js";
import PauseIcon from "../../icons/dist/react/default/pause.js";
import PipEnterIcon from "../../icons/dist/react/default/pip-enter.js";
import PipExitIcon from "../../icons/dist/react/default/pip-exit.js";
import PlayIcon from "../../icons/dist/react/default/play.js";
import RestartIcon from "../../icons/dist/react/default/restart.js";
import SeekIcon from "../../icons/dist/react/default/seek.js";
import SpinnerIcon from "../../icons/dist/react/default/spinner.js";
import VolumeHighIcon from "../../icons/dist/react/default/volume-high.js";
import VolumeLowIcon from "../../icons/dist/react/default/volume-low.js";
import VolumeOffIcon from "../../icons/dist/react/default/volume-off.js";
import { ErrorDialog } from "./error-dialog.js";
import { forwardRef } from "react";
import { isString } from "@videojs/utils/predicate";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { cn } from "@videojs/utils/style";

//#region src/presets/video/skin.tsx
const SEEK_TIME = 10;
const Button = forwardRef(function Button({ className, ...props }, ref) {
	return /* @__PURE__ */ jsx("button", {
		ref,
		type: "button",
		className: cn("media-button", className),
		...props
	});
});
const errorClasses = {
	root: "media-error",
	dialog: "media-error__dialog media-surface",
	content: "media-error__content",
	title: "media-error__title",
	description: "media-error__description",
	actions: "media-error__actions",
	close: "media-button"
};
function PlayLabel() {
	const paused = usePlayer((s) => Boolean(s.paused));
	if (usePlayer((s) => Boolean(s.ended))) return /* @__PURE__ */ jsx(Fragment, { children: "Replay" });
	return paused ? /* @__PURE__ */ jsx(Fragment, { children: "Play" }) : /* @__PURE__ */ jsx(Fragment, { children: "Pause" });
}
function CaptionsLabel() {
	return usePlayer((s) => Boolean(s.subtitlesShowing)) ? /* @__PURE__ */ jsx(Fragment, { children: "Disable captions" }) : /* @__PURE__ */ jsx(Fragment, { children: "Enable captions" });
}
function PiPLabel() {
	return usePlayer((s) => Boolean(s.pip)) ? /* @__PURE__ */ jsx(Fragment, { children: "Exit picture-in-picture" }) : /* @__PURE__ */ jsx(Fragment, { children: "Enter picture-in-picture" });
}
function FullscreenLabel() {
	return usePlayer((s) => Boolean(s.fullscreen)) ? /* @__PURE__ */ jsx(Fragment, { children: "Exit fullscreen" }) : /* @__PURE__ */ jsx(Fragment, { children: "Enter fullscreen" });
}
function VolumePopover() {
	const volumeUnsupported = usePlayer((s) => s.volumeAvailability === "unsupported");
	const muteButton = /* @__PURE__ */ jsx(MuteButton, { render: (props) => /* @__PURE__ */ jsxs(Button, {
		...props,
		className: "media-button--icon media-button--mute",
		children: [
			/* @__PURE__ */ jsx(VolumeOffIcon, { className: "media-icon media-icon--volume-off" }),
			/* @__PURE__ */ jsx(VolumeLowIcon, { className: "media-icon media-icon--volume-low" }),
			/* @__PURE__ */ jsx(VolumeHighIcon, { className: "media-icon media-icon--volume-high" })
		]
	}) });
	if (volumeUnsupported) return muteButton;
	return /* @__PURE__ */ jsxs(PopoverRoot, {
		openOnHover: true,
		delay: 200,
		closeDelay: 100,
		side: "top",
		children: [/* @__PURE__ */ jsx(PopoverTrigger, { render: muteButton }), /* @__PURE__ */ jsx(PopoverPopup, {
			className: "media-surface media-popover media-popover--volume",
			children: /* @__PURE__ */ jsxs(VolumeSliderRoot, {
				className: "media-slider",
				orientation: "vertical",
				thumbAlignment: "edge",
				children: [/* @__PURE__ */ jsx(SliderTrack, {
					className: "media-slider__track",
					children: /* @__PURE__ */ jsx(SliderFill, { className: "media-slider__fill" })
				}), /* @__PURE__ */ jsx(SliderThumb, { className: "media-slider__thumb media-slider__thumb--persistent" })]
			})
		})]
	});
}
function VideoSkin(props) {
	const { children, className, poster, ...rest } = props;
	return /* @__PURE__ */ jsxs(Container, {
		className: cn("media-default-skin media-default-skin--video", className),
		...rest,
		children: [
			children,
			poster && /* @__PURE__ */ jsx(Poster, {
				src: isString(poster) ? poster : void 0,
				render: isRenderProp(poster) ? poster : void 0
			}),
			/* @__PURE__ */ jsx(BufferingIndicator, { render: (props) => /* @__PURE__ */ jsx("div", {
				...props,
				className: "media-buffering-indicator",
				children: /* @__PURE__ */ jsx("div", {
					className: "media-surface",
					children: /* @__PURE__ */ jsx(SpinnerIcon, { className: "media-icon" })
				})
			}) }),
			/* @__PURE__ */ jsx(ErrorDialog, { classes: errorClasses }),
			/* @__PURE__ */ jsx(ControlsRoot, {
				className: "media-surface media-controls",
				children: /* @__PURE__ */ jsxs(TooltipProvider, { children: [
					/* @__PURE__ */ jsxs(TooltipRoot, {
						side: "top",
						children: [/* @__PURE__ */ jsx(TooltipTrigger, { render: /* @__PURE__ */ jsx(PlayButton, { render: (props) => /* @__PURE__ */ jsxs(Button, {
							...props,
							className: "media-button--icon media-button--play",
							children: [
								/* @__PURE__ */ jsx(RestartIcon, { className: "media-icon media-icon--restart" }),
								/* @__PURE__ */ jsx(PlayIcon, { className: "media-icon media-icon--play" }),
								/* @__PURE__ */ jsx(PauseIcon, { className: "media-icon media-icon--pause" })
							]
						}) }) }), /* @__PURE__ */ jsx(TooltipPopup, {
							className: "media-surface media-tooltip",
							children: /* @__PURE__ */ jsx(PlayLabel, {})
						})]
					}),
					/* @__PURE__ */ jsxs(TooltipRoot, {
						side: "top",
						children: [/* @__PURE__ */ jsx(TooltipTrigger, { render: /* @__PURE__ */ jsx(SeekButton, {
							seconds: -SEEK_TIME,
							render: (props) => /* @__PURE__ */ jsx(Button, {
								...props,
								className: "media-button--icon media-button--seek",
								children: /* @__PURE__ */ jsxs("span", {
									className: "media-icon__container",
									children: [/* @__PURE__ */ jsx(SeekIcon, { className: "media-icon media-icon--seek media-icon--flipped" }), /* @__PURE__ */ jsx("span", {
										className: "media-icon__label",
										children: SEEK_TIME
									})]
								})
							})
						}) }), /* @__PURE__ */ jsxs(TooltipPopup, {
							className: "media-surface media-tooltip",
							children: [
								"Seek backward ",
								SEEK_TIME,
								" seconds"
							]
						})]
					}),
					/* @__PURE__ */ jsxs(TooltipRoot, {
						side: "top",
						children: [/* @__PURE__ */ jsx(TooltipTrigger, { render: /* @__PURE__ */ jsx(SeekButton, {
							seconds: SEEK_TIME,
							render: (props) => /* @__PURE__ */ jsx(Button, {
								...props,
								className: "media-button--icon media-button--seek",
								children: /* @__PURE__ */ jsxs("span", {
									className: "media-icon__container",
									children: [/* @__PURE__ */ jsx(SeekIcon, { className: "media-icon media-icon--seek" }), /* @__PURE__ */ jsx("span", {
										className: "media-icon__label",
										children: SEEK_TIME
									})]
								})
							})
						}) }), /* @__PURE__ */ jsxs(TooltipPopup, {
							className: "media-surface media-tooltip",
							children: [
								"Seek forward ",
								SEEK_TIME,
								" seconds"
							]
						})]
					}),
					/* @__PURE__ */ jsxs(Group, {
						className: "media-time",
						children: [
							/* @__PURE__ */ jsx(Value, {
								type: "current",
								className: "media-time__value"
							}),
							/* @__PURE__ */ jsxs(TimeSliderRoot, {
								className: "media-slider",
								children: [
									/* @__PURE__ */ jsxs(SliderTrack, {
										className: "media-slider__track",
										children: [/* @__PURE__ */ jsx(SliderFill, { className: "media-slider__fill" }), /* @__PURE__ */ jsx(SliderBuffer, { className: "media-slider__buffer" })]
									}),
									/* @__PURE__ */ jsx(SliderThumb, { className: "media-slider__thumb" }),
									/* @__PURE__ */ jsxs("div", {
										className: "media-surface media-preview media-slider__preview",
										children: [
											/* @__PURE__ */ jsx(SliderThumbnail, { className: "media-preview__thumbnail" }),
											/* @__PURE__ */ jsx(SliderValue, {
												type: "pointer",
												className: "media-preview__timestamp"
											}),
											/* @__PURE__ */ jsx(SpinnerIcon, { className: "media-preview__spinner media-icon" })
										]
									})
								]
							}),
							/* @__PURE__ */ jsx(Value, {
								type: "duration",
								className: "media-time__value"
							})
						]
					}),
					/* @__PURE__ */ jsxs(TooltipRoot, {
						side: "top",
						children: [/* @__PURE__ */ jsx(TooltipTrigger, { render: /* @__PURE__ */ jsx(PlaybackRateButton, { render: (props) => /* @__PURE__ */ jsx(Button, {
							...props,
							className: "media-button--icon media-button--playback-rate"
						}) }) }), /* @__PURE__ */ jsx(TooltipPopup, {
							className: "media-surface media-tooltip",
							children: "Toggle playback rate"
						})]
					}),
					/* @__PURE__ */ jsx(VolumePopover, {}),
					/* @__PURE__ */ jsxs(TooltipRoot, {
						side: "top",
						children: [/* @__PURE__ */ jsx(TooltipTrigger, { render: /* @__PURE__ */ jsx(CaptionsButton, { render: (props) => /* @__PURE__ */ jsxs(Button, {
							...props,
							className: "media-button--icon media-button--captions",
							children: [/* @__PURE__ */ jsx(CaptionsOffIcon, { className: "media-icon media-icon--captions-off" }), /* @__PURE__ */ jsx(CaptionsOnIcon, { className: "media-icon media-icon--captions-on" })]
						}) }) }), /* @__PURE__ */ jsx(TooltipPopup, {
							className: "media-surface media-tooltip",
							children: /* @__PURE__ */ jsx(CaptionsLabel, {})
						})]
					}),
					/* @__PURE__ */ jsxs(TooltipRoot, {
						side: "top",
						children: [/* @__PURE__ */ jsx(TooltipTrigger, { render: /* @__PURE__ */ jsx(PiPButton, { render: (props) => /* @__PURE__ */ jsxs(Button, {
							...props,
							className: "media-button--icon media-button--pip",
							children: [/* @__PURE__ */ jsx(PipEnterIcon, { className: "media-icon media-icon--pip-enter" }), /* @__PURE__ */ jsx(PipExitIcon, { className: "media-icon media-icon--pip-exit" })]
						}) }) }), /* @__PURE__ */ jsx(TooltipPopup, {
							className: "media-surface media-tooltip",
							children: /* @__PURE__ */ jsx(PiPLabel, {})
						})]
					}),
					/* @__PURE__ */ jsxs(TooltipRoot, {
						side: "top",
						children: [/* @__PURE__ */ jsx(TooltipTrigger, { render: /* @__PURE__ */ jsx(FullscreenButton, { render: (props) => /* @__PURE__ */ jsxs(Button, {
							...props,
							className: "media-button--icon media-button--fullscreen",
							children: [/* @__PURE__ */ jsx(FullscreenEnterIcon, { className: "media-icon media-icon--fullscreen-enter" }), /* @__PURE__ */ jsx(FullscreenExitIcon, { className: "media-icon media-icon--fullscreen-exit" })]
						}) }) }), /* @__PURE__ */ jsx(TooltipPopup, {
							className: "media-surface media-tooltip",
							children: /* @__PURE__ */ jsx(FullscreenLabel, {})
						})]
					})
				] })
			}),
			/* @__PURE__ */ jsx("div", { className: "media-overlay" })
		]
	});
}

//#endregion
export { VideoSkin };
