// @videojs/core/dist/default/dom/presentation/fullscreen.js
import { isFunction } from "@videojs/utils/predicate";

//#region src/dom/presentation/fullscreen.ts
/** Check if the Fullscreen API is supported on this platform. */
function isFullscreenEnabled() {
	const doc = document;
	if (doc.fullscreenEnabled || doc.webkitFullscreenEnabled) return true;
	return isFunction(document.createElement("video").webkitEnterFullscreen);
}
/** Get the current fullscreen element from the document. */
function getFullscreenElement() {
	const doc = document;
	return doc.fullscreenElement ?? doc.webkitFullscreenElement ?? null;
}
/**
* Check if a specific element (or its media) is currently in fullscreen.
*
* Uses `:fullscreen` pseudo-class which works across Shadow DOM boundaries.
*/
function isFullscreenElement(container, media) {
	const video = media;
	if (video.webkitDisplayingFullscreen && video.webkitPresentationMode === "fullscreen") return true;
	const target = container ?? media;
	if (getFullscreenElement() === target) return true;
	try {
		return target.matches(":fullscreen");
	} catch {
		return false;
	}
}
/**
* Request fullscreen mode.
*
* Tries container first (to show custom UI), falls back to media element
* for platforms that only support video fullscreen (iOS Safari).
*/
async function requestFullscreen(container, media) {
	const doc = document;
	const video = media;
	if (container && (doc.fullscreenEnabled || doc.webkitFullscreenEnabled)) {
		const el = container;
		if (isFunction(el.requestFullscreen)) return el.requestFullscreen();
		if (isFunction(el.webkitRequestFullscreen)) return el.webkitRequestFullscreen();
	}
	if (isFunction(video.webkitEnterFullscreen)) {
		video.webkitEnterFullscreen();
		return;
	}
	if (isFunction(media.requestFullscreen)) return media.requestFullscreen();
	throw new DOMException("Fullscreen not supported", "NotSupportedError");
}
/** Exit fullscreen mode. */
async function exitFullscreen(media) {
	const doc = document;
	if (media) {
		const video = media;
		if (isFunction(video.webkitExitFullscreen) && video.webkitDisplayingFullscreen) {
			video.webkitExitFullscreen();
			return;
		}
	}
	if (isFunction(doc.exitFullscreen)) return doc.exitFullscreen();
	if (isFunction(doc.webkitExitFullscreen)) return doc.webkitExitFullscreen();
}

//#endregion
export { exitFullscreen, isFullscreenElement, isFullscreenEnabled, requestFullscreen };
