// @videojs/react/dist/default/index.js
'use client';

import { Container, useContainer, useContainerAttach, useMedia, useMediaAttach, usePlayer, usePlayerContext } from "./player/context.js";
import { createPlayer } from "./player/create-player.js";
import { useAlertDialogContext } from "./ui/alert-dialog/context.js";
import { mergeProps } from "./utils/merge-props.js";
import { renderElement } from "./utils/use-render.js";
import { index_parts_exports } from "./ui/alert-dialog/index.parts.js";
import { BufferingIndicator } from "./ui/buffering-indicator/buffering-indicator.js";
import { useButton } from "./ui/hooks/use-button.js";
import { CaptionsButton } from "./ui/captions-button/captions-button.js";
import { index_parts_exports as index_parts_exports$1 } from "./ui/controls/index.parts.js";
import { FullscreenButton } from "./ui/fullscreen-button/fullscreen-button.js";
import { useSlider } from "./ui/hooks/use-slider.js";
import { MuteButton } from "./ui/mute-button/mute-button.js";
import { PiPButton } from "./ui/pip-button/pip-button.js";
import { PlayButton } from "./ui/play-button/play-button.js";
import { PlaybackRateButton } from "./ui/playback-rate-button/playback-rate-button.js";
import { usePopoverContext } from "./ui/popover/context.js";
import { index_parts_exports as index_parts_exports$2 } from "./ui/popover/index.parts.js";
import { Poster } from "./ui/poster/poster.js";
import { SeekButton } from "./ui/seek-button/seek-button.js";
import { Thumbnail } from "./ui/thumbnail/thumbnail.js";
import { index_parts_exports as index_parts_exports$3 } from "./ui/slider/index.parts.js";
import { index_parts_exports as index_parts_exports$4 } from "./ui/time/index.parts.js";
import { index_parts_exports as index_parts_exports$5 } from "./ui/time-slider/index.parts.js";
import { useTooltipContext } from "./ui/tooltip/context.js";
import { index_parts_exports as index_parts_exports$6 } from "./ui/tooltip/index.parts.js";
import { index_parts_exports as index_parts_exports$7 } from "./ui/volume-slider/index.parts.js";
import { createSelector, shallowEqual } from "@videojs/store";
import { useSelector, useStore } from "@videojs/store/react";

export * from "@videojs/core/dom"

export { index_parts_exports as AlertDialog, BufferingIndicator, CaptionsButton, Container, index_parts_exports$1 as Controls, FullscreenButton, MuteButton, PiPButton, PlayButton, PlaybackRateButton, index_parts_exports$2 as Popover, Poster, SeekButton, index_parts_exports$3 as Slider, Thumbnail, index_parts_exports$4 as Time, index_parts_exports$5 as TimeSlider, index_parts_exports$6 as Tooltip, index_parts_exports$7 as VolumeSlider, createPlayer, createSelector, mergeProps, renderElement, shallowEqual, useAlertDialogContext, useButton, useContainer, useContainerAttach, useMedia, useMediaAttach, usePlayer, usePlayerContext, usePopoverContext, useSelector, useSlider, useStore, useTooltipContext };
