// @videojs/core/dist/dev/core/media/delegate.js
import { defineClassPropHooks } from "../utils/define-class-prop-hooks.js";

//#region src/core/media/delegate.ts
/**
* Mixin that intercepts `get`, `set`, and `call` to delegate property access
* and method calls to an instance of `DelegateClass` before falling through
* to the base class implementation.
*
* Works with both `CustomMediaMixin` and `ProxyMixin`.
*/
function DelegateMixin(BaseClass, DelegateClass) {
	class DelegateImpl extends BaseClass {
		#delegate = new DelegateClass();
		get(prop) {
			if (prop in this.#delegate) return this.#delegate[prop];
			return super.get?.(prop);
		}
		set(prop, val) {
			if (prop in this.#delegate) {
				this.#delegate[prop] = val;
				return;
			}
			super.set?.(prop, val);
		}
		call(prop, ...args) {
			if (prop in this.#delegate) return this.#delegate[prop](...args);
			return super.call?.(prop, ...args);
		}
		attach(target) {
			super.attach?.(target);
			this.#delegate.attach?.(target);
		}
		detach() {
			this.#delegate.detach?.();
			super.detach?.();
		}
	}
	for (let proto = DelegateClass.prototype; proto && proto !== Object.prototype; proto = Object.getPrototypeOf(proto)) defineClassPropHooks(DelegateImpl, proto);
	return DelegateImpl;
}

//#endregion
export { DelegateMixin };
