// @videojs/core/dist/dev/core/media/proxy.js
import { defineClassPropHooks } from "../utils/define-class-prop-hooks.js";

//#region src/core/media/proxy.ts
/**
* This mixin creates an API from the passed classes and proxies the methods and properties to the attached target.
*
* Many methods and properties will need no translation and are proxied directly to the attached target.
* For example, the `play` and `pause` methods are proxied directly to the attached target.
*
* Child classes can override the proxied methods and properties to provide custom behavior.
* For example, the `src` property for HLS media is proxied to the HLS engine, not the target itself.
*
* The `get`, `set`, and `call` methods can be overridden to provide catch-all custom behavior.
*/
const ProxyMixin = (PrimaryClass, ...AdditionalClasses) => {
	class MediaProxy {
		#target = null;
		get target() {
			return this.#target;
		}
		get(prop) {
			return this.target?.[prop];
		}
		set(prop, val) {
			if (this.target) this.target[prop] = val;
		}
		call(prop, ...args) {
			return (this.target?.[prop])?.apply(this.target, args);
		}
		attach(target) {
			if (!target || this.#target === target) return;
			this.#target = target;
		}
		detach() {
			if (!this.#target) return;
			this.#target = null;
		}
	}
	for (const Class of [PrimaryClass, ...AdditionalClasses]) defineClassPropHooks(MediaProxy, Class.prototype);
	return MediaProxy;
};

//#endregion
export { ProxyMixin };
