// @videojs/react/dist/dev/presets/audio/minimal-skin.js
import { Container, usePlayer } from "../../player/context.js";
import { MuteButton } from "../../ui/mute-button/mute-button.js";
import { PlayButton } from "../../ui/play-button/play-button.js";
import { PlaybackRateButton } from "../../ui/playback-rate-button/playback-rate-button.js";
import { PopoverPopup } from "../../ui/popover/popover-popup.js";
import { PopoverRoot } from "../../ui/popover/popover-root.js";
import { PopoverTrigger } from "../../ui/popover/popover-trigger.js";
import { SeekButton } from "../../ui/seek-button/seek-button.js";
import { SliderBuffer } from "../../ui/slider/slider-buffer.js";
import { SliderFill } from "../../ui/slider/slider-fill.js";
import { SliderThumb } from "../../ui/slider/slider-thumb.js";
import { SliderTrack } from "../../ui/slider/slider-track.js";
import { Group } from "../../ui/time/time-group.js";
import { Separator } from "../../ui/time/time-separator.js";
import { Value } from "../../ui/time/time-value.js";
import { TimeSliderRoot } from "../../ui/time-slider/time-slider-root.js";
import { TooltipPopup } from "../../ui/tooltip/tooltip-popup.js";
import { TooltipProvider } from "../../ui/tooltip/tooltip-provider.js";
import { TooltipRoot } from "../../ui/tooltip/tooltip-root.js";
import { TooltipTrigger } from "../../ui/tooltip/tooltip-trigger.js";
import { VolumeSliderRoot } from "../../ui/volume-slider/volume-slider-root.js";
import PauseIcon from "../../icons/dist/react/minimal/pause.js";
import PlayIcon from "../../icons/dist/react/minimal/play.js";
import RestartIcon from "../../icons/dist/react/minimal/restart.js";
import SeekIcon from "../../icons/dist/react/minimal/seek.js";
import VolumeHighIcon from "../../icons/dist/react/minimal/volume-high.js";
import VolumeLowIcon from "../../icons/dist/react/minimal/volume-low.js";
import VolumeOffIcon from "../../icons/dist/react/minimal/volume-off.js";
import { forwardRef } from "react";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { cn } from "@videojs/utils/style";

//#region src/presets/audio/minimal-skin.tsx
const SEEK_TIME = 10;
const Button = forwardRef(function Button({ className, ...props }, ref) {
	return /* @__PURE__ */ jsx("button", {
		ref,
		type: "button",
		className: cn("media-button", className),
		...props
	});
});
function PlayLabel() {
	const paused = usePlayer((s) => Boolean(s.paused));
	if (usePlayer((s) => Boolean(s.ended))) return /* @__PURE__ */ jsx(Fragment, { children: "Replay" });
	return paused ? /* @__PURE__ */ jsx(Fragment, { children: "Play" }) : /* @__PURE__ */ jsx(Fragment, { children: "Pause" });
}
function VolumePopover() {
	const volumeUnsupported = usePlayer((s) => s.volumeAvailability === "unsupported");
	const muteButton = /* @__PURE__ */ jsx(MuteButton, { render: (props) => /* @__PURE__ */ jsxs(Button, {
		...props,
		className: "media-button--icon media-button--mute",
		children: [
			/* @__PURE__ */ jsx(VolumeOffIcon, { className: "media-icon media-icon--volume-off" }),
			/* @__PURE__ */ jsx(VolumeLowIcon, { className: "media-icon media-icon--volume-low" }),
			/* @__PURE__ */ jsx(VolumeHighIcon, { className: "media-icon media-icon--volume-high" })
		]
	}) });
	if (volumeUnsupported) return muteButton;
	return /* @__PURE__ */ jsxs(PopoverRoot, {
		openOnHover: true,
		delay: 200,
		closeDelay: 100,
		side: "left",
		children: [/* @__PURE__ */ jsx(PopoverTrigger, { render: muteButton }), /* @__PURE__ */ jsx(PopoverPopup, {
			className: "media-popover media-popover--volume",
			children: /* @__PURE__ */ jsxs(VolumeSliderRoot, {
				className: "media-slider",
				orientation: "horizontal",
				thumbAlignment: "edge",
				children: [/* @__PURE__ */ jsx(SliderTrack, {
					className: "media-slider__track",
					children: /* @__PURE__ */ jsx(SliderFill, { className: "media-slider__fill" })
				}), /* @__PURE__ */ jsx(SliderThumb, { className: "media-slider__thumb media-slider__thumb--persistent" })]
			})
		})]
	});
}
function MinimalAudioSkin(props) {
	const { children, className, ...rest } = props;
	return /* @__PURE__ */ jsxs(Container, {
		className: cn("media-minimal-skin media-minimal-skin--audio", className),
		...rest,
		children: [children, /* @__PURE__ */ jsx("div", {
			className: "media-controls",
			children: /* @__PURE__ */ jsxs(TooltipProvider, { children: [
				/* @__PURE__ */ jsxs("div", {
					className: "media-button-group",
					children: [
						/* @__PURE__ */ jsxs(TooltipRoot, {
							side: "top",
							children: [/* @__PURE__ */ jsx(TooltipTrigger, { render: /* @__PURE__ */ jsx(PlayButton, { render: (props) => /* @__PURE__ */ jsxs(Button, {
								...props,
								className: "media-button--icon media-button--play",
								children: [
									/* @__PURE__ */ jsx(RestartIcon, { className: "media-icon media-icon--restart" }),
									/* @__PURE__ */ jsx(PlayIcon, { className: "media-icon media-icon--play" }),
									/* @__PURE__ */ jsx(PauseIcon, { className: "media-icon media-icon--pause" })
								]
							}) }) }), /* @__PURE__ */ jsx(TooltipPopup, {
								className: "media-tooltip",
								children: /* @__PURE__ */ jsx(PlayLabel, {})
							})]
						}),
						/* @__PURE__ */ jsxs(TooltipRoot, {
							side: "top",
							children: [/* @__PURE__ */ jsx(TooltipTrigger, { render: /* @__PURE__ */ jsx(SeekButton, {
								seconds: -SEEK_TIME,
								render: (props) => /* @__PURE__ */ jsx(Button, {
									...props,
									className: "media-button--icon media-button--seek",
									children: /* @__PURE__ */ jsxs("span", {
										className: "media-icon__container",
										children: [/* @__PURE__ */ jsx(SeekIcon, { className: "media-icon media-icon--seek media-icon--flipped" }), /* @__PURE__ */ jsx("span", {
											className: "media-icon__label",
											children: SEEK_TIME
										})]
									})
								})
							}) }), /* @__PURE__ */ jsxs(TooltipPopup, {
								className: "media-tooltip",
								children: [
									"Seek backward ",
									SEEK_TIME,
									" seconds"
								]
							})]
						}),
						/* @__PURE__ */ jsxs(TooltipRoot, {
							side: "top",
							children: [/* @__PURE__ */ jsx(TooltipTrigger, { render: /* @__PURE__ */ jsx(SeekButton, {
								seconds: SEEK_TIME,
								render: (props) => /* @__PURE__ */ jsx(Button, {
									...props,
									className: "media-button--icon media-button--seek",
									children: /* @__PURE__ */ jsxs("span", {
										className: "media-icon__container",
										children: [/* @__PURE__ */ jsx(SeekIcon, { className: "media-icon media-icon--seek" }), /* @__PURE__ */ jsx("span", {
											className: "media-icon__label",
											children: SEEK_TIME
										})]
									})
								})
							}) }), /* @__PURE__ */ jsxs(TooltipPopup, {
								className: "media-tooltip",
								children: [
									"Seek forward ",
									SEEK_TIME,
									" seconds"
								]
							})]
						})
					]
				}),
				/* @__PURE__ */ jsxs("div", {
					className: "media-time-controls",
					children: [/* @__PURE__ */ jsxs(Group, {
						className: "media-time",
						children: [
							/* @__PURE__ */ jsx(Value, {
								type: "current",
								className: "media-time__value media-time__value--current"
							}),
							/* @__PURE__ */ jsx(Separator, { className: "media-time__separator" }),
							/* @__PURE__ */ jsx(Value, {
								type: "duration",
								className: "media-time__value media-time__value--duration"
							})
						]
					}), /* @__PURE__ */ jsxs(TimeSliderRoot, {
						className: "media-slider",
						children: [/* @__PURE__ */ jsxs(SliderTrack, {
							className: "media-slider__track",
							children: [/* @__PURE__ */ jsx(SliderFill, { className: "media-slider__fill" }), /* @__PURE__ */ jsx(SliderBuffer, { className: "media-slider__buffer" })]
						}), /* @__PURE__ */ jsx(SliderThumb, { className: "media-slider__thumb" })]
					})]
				}),
				/* @__PURE__ */ jsxs("div", {
					className: "media-button-group",
					children: [/* @__PURE__ */ jsxs(TooltipRoot, {
						side: "top",
						children: [/* @__PURE__ */ jsx(TooltipTrigger, { render: /* @__PURE__ */ jsx(PlaybackRateButton, { render: (props) => /* @__PURE__ */ jsx(Button, {
							...props,
							className: "media-button--icon media-button--playback-rate"
						}) }) }), /* @__PURE__ */ jsx(TooltipPopup, {
							className: "media-tooltip",
							children: "Toggle playback rate"
						})]
					}), /* @__PURE__ */ jsx(VolumePopover, {})]
				})
			] })
		})]
	});
}

//#endregion
export { MinimalAudioSkin };
