// @videojs/react/dist/dev/presets/audio/skin.tailwind.js
import { Container, usePlayer } from "../../player/context.js";
import { MuteButton } from "../../ui/mute-button/mute-button.js";
import { PlayButton } from "../../ui/play-button/play-button.js";
import { PlaybackRateButton } from "../../ui/playback-rate-button/playback-rate-button.js";
import { PopoverPopup } from "../../ui/popover/popover-popup.js";
import { PopoverRoot } from "../../ui/popover/popover-root.js";
import { PopoverTrigger } from "../../ui/popover/popover-trigger.js";
import { SeekButton } from "../../ui/seek-button/seek-button.js";
import { SliderBuffer } from "../../ui/slider/slider-buffer.js";
import { SliderFill as SliderFill$1 } from "../../ui/slider/slider-fill.js";
import { SliderThumb as SliderThumb$1 } from "../../ui/slider/slider-thumb.js";
import { SliderTrack as SliderTrack$1 } from "../../ui/slider/slider-track.js";
import { Group } from "../../ui/time/time-group.js";
import { Value } from "../../ui/time/time-value.js";
import { TimeSliderRoot } from "../../ui/time-slider/time-slider-root.js";
import { TooltipPopup } from "../../ui/tooltip/tooltip-popup.js";
import { TooltipProvider } from "../../ui/tooltip/tooltip-provider.js";
import { TooltipRoot } from "../../ui/tooltip/tooltip-root.js";
import { TooltipTrigger } from "../../ui/tooltip/tooltip-trigger.js";
import { VolumeSliderRoot } from "../../ui/volume-slider/volume-slider-root.js";
import { iconState } from "../../skins/dist/default/shared/tailwind/icon-state.js";
import { button } from "../../skins/dist/default/default/tailwind/components/button.js";
import { icon, iconContainer, iconFlipped } from "../../skins/dist/default/default/tailwind/components/icon.js";
import { playbackRate } from "../../skins/dist/default/default/tailwind/components/playback-rate.js";
import { seek } from "../../skins/dist/default/default/tailwind/components/seek.js";
import { time } from "../../skins/dist/default/default/tailwind/components/time.js";
import { controls, popup, root, slider } from "../../skins/dist/default/default/tailwind/audio.tailwind.js";
import PauseIcon from "../../icons/dist/react/default/pause.js";
import PlayIcon from "../../icons/dist/react/default/play.js";
import RestartIcon from "../../icons/dist/react/default/restart.js";
import SeekIcon from "../../icons/dist/react/default/seek.js";
import VolumeHighIcon from "../../icons/dist/react/default/volume-high.js";
import VolumeLowIcon from "../../icons/dist/react/default/volume-low.js";
import VolumeOffIcon from "../../icons/dist/react/default/volume-off.js";
import { forwardRef } from "react";
import { Fragment, jsx, jsxs } from "react/jsx-runtime";
import { cn } from "@videojs/utils/style";

//#region src/presets/audio/skin.tailwind.tsx
const SEEK_TIME = 10;
const Button = forwardRef(function Button({ className, variant, ...props }, ref) {
	return /* @__PURE__ */ jsx("button", {
		ref,
		type: "button",
		className: cn(button.base, variant === "icon" ? button.icon : button.default, className),
		...props
	});
});
const SliderRoot = forwardRef(function SliderRoot({ className, ...props }, ref) {
	return /* @__PURE__ */ jsx("div", {
		ref,
		className: cn(slider.root, className),
		...props
	});
});
const SliderTrack = forwardRef(function SliderTrack({ className, ...props }, ref) {
	return /* @__PURE__ */ jsx("div", {
		ref,
		className: cn(slider.track, className),
		...props
	});
});
const SliderFill = forwardRef(function SliderFill({ type = "fill", className, ...props }, ref) {
	return /* @__PURE__ */ jsx("div", {
		ref,
		className: cn(slider.fill.base, type === "fill" ? slider.fill.fill : slider.fill.buffer, className),
		...props
	});
});
const SliderThumb = forwardRef(function SliderThumb({ persistent, className, ...props }, ref) {
	return /* @__PURE__ */ jsx("div", {
		ref,
		className: cn(slider.thumb.base, persistent ? slider.thumb.persistent : slider.thumb.interactive, className),
		...props
	});
});
function PlayLabel() {
	const paused = usePlayer((s) => Boolean(s.paused));
	if (usePlayer((s) => Boolean(s.ended))) return /* @__PURE__ */ jsx(Fragment, { children: "Replay" });
	return paused ? /* @__PURE__ */ jsx(Fragment, { children: "Play" }) : /* @__PURE__ */ jsx(Fragment, { children: "Pause" });
}
function VolumePopover() {
	const volumeUnsupported = usePlayer((s) => s.volumeAvailability === "unsupported");
	const muteButton = /* @__PURE__ */ jsx(MuteButton, { render: (props) => /* @__PURE__ */ jsxs(Button, {
		variant: "icon",
		...props,
		className: iconState.mute.button,
		children: [
			/* @__PURE__ */ jsx(VolumeOffIcon, { className: cn(icon, iconState.mute.volumeOff) }),
			/* @__PURE__ */ jsx(VolumeLowIcon, { className: cn(icon, iconState.mute.volumeLow) }),
			/* @__PURE__ */ jsx(VolumeHighIcon, { className: cn(icon, iconState.mute.volumeHigh) })
		]
	}) });
	if (volumeUnsupported) return muteButton;
	return /* @__PURE__ */ jsxs(PopoverRoot, {
		openOnHover: true,
		delay: 200,
		closeDelay: 100,
		side: "top",
		children: [/* @__PURE__ */ jsx(PopoverTrigger, { render: muteButton }), /* @__PURE__ */ jsx(PopoverPopup, {
			className: cn(popup.popover, popup.volume),
			children: /* @__PURE__ */ jsxs(VolumeSliderRoot, {
				orientation: "vertical",
				thumbAlignment: "edge",
				render: (props) => /* @__PURE__ */ jsx(SliderRoot, { ...props }),
				children: [/* @__PURE__ */ jsx(SliderTrack$1, {
					render: (props) => /* @__PURE__ */ jsx(SliderTrack, { ...props }),
					children: /* @__PURE__ */ jsx(SliderFill$1, { render: (props) => /* @__PURE__ */ jsx(SliderFill, { ...props }) })
				}), /* @__PURE__ */ jsx(SliderThumb$1, { render: (props) => /* @__PURE__ */ jsx(SliderThumb, {
					persistent: true,
					...props
				}) })]
			})
		})]
	});
}
function AudioSkinTailwind(props) {
	const { children, className, ...rest } = props;
	return /* @__PURE__ */ jsxs(Container, {
		className: cn(root, className),
		...rest,
		children: [children, /* @__PURE__ */ jsx("div", {
			className: controls,
			children: /* @__PURE__ */ jsxs(TooltipProvider, { children: [
				/* @__PURE__ */ jsxs(TooltipRoot, {
					side: "top",
					children: [/* @__PURE__ */ jsx(TooltipTrigger, { render: /* @__PURE__ */ jsx(PlayButton, { render: (props) => /* @__PURE__ */ jsxs(Button, {
						variant: "icon",
						...props,
						className: iconState.play.button,
						children: [
							/* @__PURE__ */ jsx(RestartIcon, { className: cn(icon, iconState.play.restart) }),
							/* @__PURE__ */ jsx(PlayIcon, { className: cn(icon, iconState.play.play) }),
							/* @__PURE__ */ jsx(PauseIcon, { className: cn(icon, iconState.play.pause) })
						]
					}) }) }), /* @__PURE__ */ jsx(TooltipPopup, {
						className: cn(popup.tooltip),
						children: /* @__PURE__ */ jsx(PlayLabel, {})
					})]
				}),
				/* @__PURE__ */ jsxs(TooltipRoot, {
					side: "top",
					children: [/* @__PURE__ */ jsx(TooltipTrigger, { render: /* @__PURE__ */ jsx(SeekButton, {
						seconds: -SEEK_TIME,
						render: (props) => /* @__PURE__ */ jsx(Button, {
							variant: "icon",
							...props,
							className: seek.button,
							children: /* @__PURE__ */ jsxs("span", {
								className: iconContainer,
								children: [/* @__PURE__ */ jsx(SeekIcon, { className: cn(icon, iconFlipped) }), /* @__PURE__ */ jsx("span", {
									className: cn(seek.label, seek.labelBackward),
									children: SEEK_TIME
								})]
							})
						})
					}) }), /* @__PURE__ */ jsxs(TooltipPopup, {
						className: cn(popup.tooltip),
						children: [
							"Seek backward ",
							SEEK_TIME,
							" seconds"
						]
					})]
				}),
				/* @__PURE__ */ jsxs(TooltipRoot, {
					side: "top",
					children: [/* @__PURE__ */ jsx(TooltipTrigger, { render: /* @__PURE__ */ jsx(SeekButton, {
						seconds: SEEK_TIME,
						render: (props) => /* @__PURE__ */ jsx(Button, {
							variant: "icon",
							...props,
							className: seek.button,
							children: /* @__PURE__ */ jsxs("span", {
								className: iconContainer,
								children: [/* @__PURE__ */ jsx(SeekIcon, { className: icon }), /* @__PURE__ */ jsx("span", {
									className: cn(seek.label, seek.labelForward),
									children: SEEK_TIME
								})]
							})
						})
					}) }), /* @__PURE__ */ jsxs(TooltipPopup, {
						className: cn(popup.tooltip),
						children: [
							"Seek forward ",
							SEEK_TIME,
							" seconds"
						]
					})]
				}),
				/* @__PURE__ */ jsxs(Group, {
					className: time.group,
					children: [
						/* @__PURE__ */ jsx(Value, {
							type: "current",
							className: time.current
						}),
						/* @__PURE__ */ jsxs(TimeSliderRoot, {
							render: (props) => /* @__PURE__ */ jsx(SliderRoot, { ...props }),
							children: [/* @__PURE__ */ jsxs(SliderTrack$1, {
								render: (props) => /* @__PURE__ */ jsx(SliderTrack, { ...props }),
								children: [/* @__PURE__ */ jsx(SliderFill$1, { render: (props) => /* @__PURE__ */ jsx(SliderFill, { ...props }) }), /* @__PURE__ */ jsx(SliderBuffer, { render: (props) => /* @__PURE__ */ jsx(SliderFill, {
									type: "buffer",
									...props
								}) })]
							}), /* @__PURE__ */ jsx(SliderThumb$1, { render: (props) => /* @__PURE__ */ jsx(SliderThumb, { ...props }) })]
						}),
						/* @__PURE__ */ jsx(Value, {
							type: "duration",
							className: time.duration
						})
					]
				}),
				/* @__PURE__ */ jsxs(TooltipRoot, {
					side: "top",
					children: [/* @__PURE__ */ jsx(TooltipTrigger, { render: /* @__PURE__ */ jsx(PlaybackRateButton, { render: (props) => /* @__PURE__ */ jsx(Button, {
						variant: "icon",
						...props,
						className: playbackRate.button
					}) }) }), /* @__PURE__ */ jsx(TooltipPopup, {
						className: cn(popup.tooltip),
						children: "Toggle playback rate"
					})]
				}),
				/* @__PURE__ */ jsx(VolumePopover, {})
			] })
		})]
	});
}

//#endregion
export { AudioSkinTailwind };
