// @videojs/utils/dist/dom.js
import { animationFrame } from "./dom/animation-frame.js";
import { namedNodeMapToObject } from "./dom/attributes.js";
import { isRTL } from "./dom/direction.js";
import { onEvent } from "./dom/event.js";
import { supportsAnchorPositioning, supportsAnimationFrame, supportsIdleCallback } from "./dom/supports.js";
import { idleCallback } from "./dom/idle-callback.js";
import { listen } from "./dom/listen.js";
import { tryHidePopover, tryShowPopover } from "./dom/popover.js";
import { isHTMLAudioElement, isHTMLMediaElement, isHTMLVideoElement } from "./dom/predicates.js";
import { rafThrottle } from "./dom/raf-throttle.js";
import { getSlottedElement, querySlot } from "./dom/slotted.js";
import { applyStyles, resolveCSSLength } from "./dom/style.js";
import { findTrackElement, getTextTrackList } from "./dom/text-track.js";
import { serializeTimeRanges } from "./dom/time-ranges.js";

export { animationFrame, applyStyles, findTrackElement, getSlottedElement, getTextTrackList, idleCallback, isHTMLAudioElement, isHTMLMediaElement, isHTMLVideoElement, isRTL, listen, namedNodeMapToObject, onEvent, querySlot, rafThrottle, resolveCSSLength, serializeTimeRanges, supportsAnchorPositioning, supportsAnimationFrame, supportsIdleCallback, tryHidePopover, tryShowPopover };
