// @videojs/react/dist/dev/skins/dist/default/default/tailwind/audio.tailwind.js
import { iconState } from "../../shared/tailwind/icon-state.js";
import { bufferingIndicator as bufferingIndicator$1 } from "./components/buffering.js";
import { controls as controls$1 } from "./components/controls.js";
import { error as error$1 } from "./components/error.js";
import { popup as popup$1 } from "./components/popup.js";
import { root as root$1 } from "./components/root.js";
import { slider as slider$1 } from "./components/slider.js";
import { surface as surface$1 } from "./components/surface.js";
import { button } from "./components/button.js";
import { icon, iconContainer, iconFlipped } from "./components/icon.js";
import { playbackRate } from "./components/playback-rate.js";
import { seek } from "./components/seek.js";
import { time } from "./components/time.js";
import { cn } from "@videojs/utils/style";

//#region ../skins/dist/default/default/tailwind/audio.tailwind.js
const root = cn(root$1, "[--media-controls-text-color:var(--media-color-primary,oklch(0_0_0))]", "dark:[--media-controls-text-color:var(--media-color-primary,oklch(1_0_0))]");
const surface = cn(surface$1, "bg-white/50 dark:bg-black/40", "backdrop-blur-lg backdrop-saturate-150", "ring-black/5 shadow-sm shadow-black/15", "after:ring-white/10", "[@media(prefers-reduced-transparency:reduce)]:bg-white/70 dark:[@media(prefers-reduced-transparency:reduce)]:bg-black/70", "contrast-more:bg-white/90 dark:contrast-more:bg-black/90");
const controls = cn(controls$1, surface, "text-(--media-controls-text-color)");
const slider = {
	...slider$1,
	track: cn(slider$1.track, "bg-black/10", "dark:bg-white/20 dark:ring-1 dark:ring-black/5")
};
const popup = {
	...popup$1,
	popover: cn(surface, popup$1.popover),
	tooltip: cn(surface, popup$1.tooltip)
};
const bufferingIndicator = {
	...bufferingIndicator$1,
	container: cn(bufferingIndicator$1.container, surface)
};
const error = {
	...error$1,
	dialog: cn(error$1.dialog, surface)
};

//#endregion
export { controls, popup, root, slider };
