// @videojs/react/dist/default/skins/dist/default/default/tailwind/video.tailwind.js
import { iconState } from "../../shared/tailwind/icon-state.js";
import { bufferingIndicator as bufferingIndicator$1 } from "./components/buffering.js";
import { controls as controls$1 } from "./components/controls.js";
import { error as error$1 } from "./components/error.js";
import { popup as popup$1 } from "./components/popup.js";
import { root as root$1 } from "./components/root.js";
import { slider as slider$1 } from "./components/slider.js";
import { surface as surface$1 } from "./components/surface.js";
import { button } from "./components/button.js";
import { icon, iconContainer, iconFlipped } from "./components/icon.js";
import { playbackRate } from "./components/playback-rate.js";
import { seek } from "./components/seek.js";
import { time } from "./components/time.js";
import { preview as preview$1 } from "./components/preview.js";
import { overlay } from "./components/overlay.js";
import { poster } from "./components/poster.js";
import { cn } from "@videojs/utils/style";

//#region ../skins/dist/default/default/tailwind/video.tailwind.js
const root = (isShadowDOM) => cn(root$1, "bg-black", "after:absolute after:pointer-events-none after:rounded-[inherit] after:z-10", "after:inset-0 after:ring-1 after:ring-inset after:ring-black/10 dark:after:ring-white/15", {
	"[&_::slotted(video)]:block [&_::slotted(video)]:w-full [&_::slotted(video)]:h-full [&_::slotted(video)]:rounded-(--media-video-border-radius) [&_::slotted(video)]:[object-fit:var(--media-object-fit,contain)] [&_::slotted(video)]:[object-position:var(--media-object-position,center)]": isShadowDOM,
	"[&_video]:block [&_video]:w-full [&_video]:h-full [&_video]:rounded-[inherit] [&_video]:[object-fit:var(--media-object-fit,contain)] [&_video]:[object-position:var(--media-object-position,center)]": !isShadowDOM
}, "[--media-video-border-radius:var(--media-border-radius,2rem)]", "[--media-controls-transition-duration:100ms]", "[--media-controls-transition-delay:0ms]", "[@media(pointer:fine)]:has-[[data-controls]:not([data-visible])]:[--media-controls-transition-delay:500ms]", "[@media(pointer:fine)]:has-[[data-controls]:not([data-visible])]:[--media-controls-transition-duration:300ms]", "[@media(pointer:coarse)]:has-[[data-controls]:not([data-visible])]:[--media-controls-transition-duration:150ms]", "motion-reduce:has-[[data-controls]:not([data-visible])]:[--media-controls-transition-duration:100ms]", "[--media-caption-track-y:-0.5rem]", "[--media-caption-track-delay:calc(var(--media-controls-transition-delay)_+_25ms)]", "[--media-caption-track-duration:var(--media-controls-transition-duration)]", "has-[[data-controls][data-visible]]:[--media-caption-track-y:-3.5rem]", !isShadowDOM ? [
	"[&_video::-webkit-media-text-track-container]:transition-[translate]",
	"[&_video::-webkit-media-text-track-container]:duration-(--media-caption-track-duration)",
	"[&_video::-webkit-media-text-track-container]:ease-out",
	"[&_video::-webkit-media-text-track-container]:delay-(--media-caption-track-delay)",
	"[&_video::-webkit-media-text-track-container]:translate-y-(--media-caption-track-y)",
	"[&_video::-webkit-media-text-track-container]:scale-98",
	"[&_video::-webkit-media-text-track-container]:z-1",
	"[&_video::-webkit-media-text-track-container]:font-[inherit]"
] : [], "[&:fullscreen]:[--media-border-radius:0]", {
	"[&:fullscreen_video]:object-contain": !isShadowDOM,
	"[&:fullscreen_::slotted(video)]:object-contain": isShadowDOM
});
const surface = cn(surface$1, "bg-white/10", "backdrop-saturate-150 backdrop-blur-lg", "ring-black/15 shadow-black/10", "after:ring-white/5", "[@media(prefers-reduced-transparency:reduce)]:bg-black/70", "contrast-more:bg-black/90");
const controls = cn(controls$1, surface, "absolute bottom-3 inset-x-3", "[color:var(--media-color-primary,oklch(1_0_0))] z-10", "ease-out origin-bottom", "duration-(--media-controls-transition-duration)", "delay-(--media-controls-transition-delay)", "[@media(pointer:fine)]:will-change-[scale,filter,opacity]", "[@media(pointer:fine)]:transition-[scale,filter,opacity]", "[@media(pointer:coarse)]:will-change-[scale,opacity]", "[@media(pointer:coarse)]:transition-[scale,opacity]", "not-data-visible:pointer-events-none not-data-visible:opacity-0", "motion-safe:not-data-visible:scale-90", "[@media(pointer:fine)]:motion-safe:not-data-visible:blur-sm");
const preview = {
	...preview$1,
	root: cn("absolute left-(--media-slider-pointer) bottom-[calc(100%+1.2rem)] -translate-x-1/2", "opacity-0 scale-80 blur-sm origin-bottom", "transition-[scale,opacity,filter] duration-150", "group-data-pointing/slider:opacity-100 group-data-pointing/slider:scale-100 group-data-pointing/slider:blur-none", "[&:has([role=img][data-hidden])]:opacity-0 [&:has([role=img][data-hidden])]:scale-80 [&:has([role=img][data-hidden])]:blur-sm", "[&:has([role=img][data-loading])]:max-h-24", surface, preview$1.root),
	thumbnail: cn(preview$1.thumbnail, "max-w-44")
};
const slider = {
	...slider$1,
	track: cn(slider$1.track, "bg-white/20 shadow-[0_0_0_1px_oklch(0_0_0/0.05)]")
};
const popup = {
	...popup$1,
	popover: cn(surface, popup$1.popover),
	tooltip: cn(surface, popup$1.tooltip)
};
const bufferingIndicator = {
	...bufferingIndicator$1,
	container: cn(bufferingIndicator$1.container, surface)
};
const error = {
	...error$1,
	dialog: cn(error$1.dialog, surface)
};

//#endregion
export { bufferingIndicator, controls, error, popup, preview, root, slider };
