// @videojs/react/dist/dev/ui/volume-slider/volume-slider-root.js
'use client';

import { usePlayer } from "../../player/context.js";
import { renderElement } from "../../utils/use-render.js";
import { useLatestRef } from "../../utils/use-latest-ref.js";
import { useSlider } from "../hooks/use-slider.js";
import { SliderProvider } from "../slider/context.js";
import { getSliderCSSVars, logMissingFeature, selectVolume } from "@videojs/core/dom";
import { forwardRef, useState } from "react";
import { jsx } from "react/jsx-runtime";
import { VolumeSliderCore, VolumeSliderDataAttrs } from "@videojs/core";

//#region src/ui/volume-slider/volume-slider-root.tsx
const noopVolume = {
	volume: 0,
	muted: false,
	volumeAvailability: "unsupported",
	setVolume: () => 0,
	toggleMuted: () => false
};
const VolumeSliderRoot = forwardRef(function VolumeSliderRoot(componentProps, forwardedRef) {
	const { render, className, style, label, orientation, step = VolumeSliderCore.defaultProps.step, largeStep = VolumeSliderCore.defaultProps.largeStep, disabled, thumbAlignment, onDragStart, onDragEnd, ...elementProps } = componentProps;
	const volume = usePlayer(selectVolume);
	const [core] = useState(() => new VolumeSliderCore());
	core.setProps({
		label,
		orientation,
		step,
		largeStep,
		disabled,
		thumbAlignment
	});
	const volumeRef = useLatestRef(volume);
	const { state, cssVars, rootRef, thumbRef, rootProps, rootStyle, thumbProps } = useSlider({
		computeState: (input) => {
			core.setInput(input);
			core.setMedia(volume ?? noopVolume);
			return core.getState();
		},
		getPercent: () => volume ? volume.volume * 100 : 0,
		getStepPercent: () => core.getStepPercent(),
		getLargeStepPercent: () => core.getLargeStepPercent(),
		orientation,
		disabled,
		adjustPercent: (rawPercent, thumbSize, trackSize) => core.adjustPercentForAlignment(rawPercent, thumbSize, trackSize),
		getCSSVars: getSliderCSSVars,
		onValueChange: (percent) => {
			volumeRef.current?.setVolume(percent / 100);
		},
		onValueCommit: (percent) => {
			volumeRef.current?.setVolume(percent / 100);
		},
		onDragStart,
		onDragEnd
	});
	if (!volume) {
		logMissingFeature("VolumeSlider", "volume");
		return null;
	}
	return /* @__PURE__ */ jsx(SliderProvider, {
		value: {
			state,
			pointerValue: core.valueFromPercent(state.pointerPercent),
			thumbRef,
			thumbProps,
			stateAttrMap: VolumeSliderDataAttrs,
			getAttrs: (sliderState) => core.getAttrs(sliderState),
			formatValue: (value) => `${Math.round(value)}%`
		},
		children: renderElement("div", {
			render,
			className,
			style
		}, {
			state,
			stateAttrMap: VolumeSliderDataAttrs,
			ref: [forwardedRef, rootRef],
			props: [
				{ style: {
					...cssVars,
					...rootStyle
				} },
				rootProps,
				elementProps
			]
		})
	});
});

//#endregion
export { VolumeSliderRoot };
