// @videojs/spf/dist/dom/index.js
import "../src-DRsS_H80.js";
import { a as trackPlaybackInitiated, c as destroyVttParser, i as appendSegment, l as parseVttSegment, o as trackCurrentTime, r as flushBuffer, s as loadTextTrackCues, t as SpfMedia, u as loadSegments } from "../adapter-ClnRGbQf.js";
import { listen } from "@videojs/utils/dom";

//#region src/dom/features/track-playback-rate.ts
/**
* Track playback rate from the media element.
*
* Mirrors `mediaElement.playbackRate` into reactive state on each `ratechange`
* event. Also syncs immediately when a media element becomes available.
*
* @example
* const cleanup = trackPlaybackRate({ state, owners });
*/
function trackPlaybackRate({ state, owners }) {
	let lastMediaElement;
	let removeListener = null;
	const unsubscribe = owners.subscribe((currentOwners) => {
		const { mediaElement } = currentOwners;
		if (mediaElement === lastMediaElement) return;
		removeListener?.();
		removeListener = null;
		lastMediaElement = mediaElement;
		if (!mediaElement) return;
		state.patch({ playbackRate: mediaElement.playbackRate });
		removeListener = listen(mediaElement, "ratechange", () => {
			state.patch({ playbackRate: mediaElement.playbackRate });
		});
	});
	return () => {
		removeListener?.();
		unsubscribe();
	};
}

//#endregion
export { SpfMedia, appendSegment, destroyVttParser, flushBuffer, loadSegments, loadTextTrackCues, parseVttSegment, trackCurrentTime, trackPlaybackInitiated, trackPlaybackRate };
