// @videojs/core/dist/default/dom/media/dash/index.js
import { DelegateMixin } from "../../../core/media/delegate.js";
import { CustomMediaMixin } from "../custom-media-element/index.js";
import { MediaProxyMixin } from "../proxy.js";
import * as dashjs from "dashjs";

//#region src/dom/media/dash/index.ts
var DashMediaDelegateBase = class {
	#engine;
	constructor() {
		this.#engine = dashjs.MediaPlayer().create();
		this.#engine.initialize(void 0, void 0, false);
	}
	get engine() {
		return this.#engine;
	}
	attach(target) {
		this.#engine.attachView(target);
	}
	detach() {
		this.#engine.attachView(null);
	}
	destroy() {
		this.#engine.destroy();
	}
	set src(src) {
		this.#engine.attachSource(src);
	}
	get src() {
		return this.#engine.getSource() ?? "";
	}
};
var DashCustomMedia = class extends DelegateMixin(CustomMediaMixin(globalThis.HTMLElement ?? class {}, { tag: "video" }), DashMediaDelegateBase) {};
var DashMedia = class extends DelegateMixin(MediaProxyMixin, DashMediaDelegateBase) {};

//#endregion
export { DashCustomMedia, DashMedia, DashMediaDelegateBase };
