// @videojs/core/dist/default/dom/store/features/pip.js
import { definePlayerFeature } from "../../feature.js";
import { exitFullscreen, isFullscreenElement } from "../../presentation/fullscreen.js";
import { exitPictureInPicture, isPictureInPictureElement, isPictureInPictureEnabled, requestPictureInPicture } from "../../presentation/pip.js";
import { listen } from "@videojs/utils/dom";

//#region src/dom/store/features/pip.ts
const pipFeature = definePlayerFeature({
	name: "pip",
	state: ({ target }) => ({
		pip: false,
		pipAvailability: "unavailable",
		async requestPictureInPicture() {
			const { media, container } = target();
			if (isFullscreenElement(container, media)) await exitFullscreen();
			return requestPictureInPicture(media);
		},
		async exitPictureInPicture() {
			const { media } = target();
			return exitPictureInPicture(media);
		}
	}),
	attach({ target, signal, set }) {
		const { media } = target;
		set({ pipAvailability: isPictureInPictureEnabled() ? "available" : "unsupported" });
		const sync = () => set({ pip: isPictureInPictureElement(media) });
		sync();
		listen(media, "enterpictureinpicture", sync, { signal });
		listen(media, "leavepictureinpicture", sync, { signal });
		if ("webkitPresentationMode" in media) listen(media, "webkitpresentationmodechanged", sync, { signal });
	}
});

//#endregion
export { pipFeature };
