// @videojs/core/dist/default/dom/store/features/text-track.js
import { definePlayerFeature } from "../../feature.js";
import { findTrackElement, getTextTrackList, listen } from "@videojs/utils/dom";

//#region src/dom/store/features/text-track.ts
const textTrackFeature = definePlayerFeature({
	name: "textTrack",
	state: ({ target }) => ({
		chaptersCues: [],
		thumbnailCues: [],
		thumbnailTrackSrc: null,
		textTrackList: [],
		subtitlesShowing: false,
		toggleSubtitles(forceShow) {
			const subtitlesTracks = getTextTrackList(target().media, (track) => track.kind === "subtitles" || track.kind === "captions");
			if (!subtitlesTracks.length) return false;
			const showing = subtitlesTracks.some((track) => track.mode === "showing");
			const nextShowing = forceShow ?? !showing;
			for (const track of subtitlesTracks) track.mode = nextShowing ? "showing" : "disabled";
			return nextShowing;
		}
	}),
	attach({ target, signal, set }) {
		const { media } = target;
		let trackCleanup = null;
		function sync() {
			trackCleanup?.abort();
			trackCleanup = new AbortController();
			let chaptersTrack = null;
			let thumbnailTrack = null;
			const textTrackList = [];
			let subtitlesShowing = false;
			for (let i = 0; i < media.textTracks.length; i++) {
				const track = media.textTracks[i];
				if (!chaptersTrack && track.kind === "chapters") chaptersTrack = track;
				if (!thumbnailTrack && track.kind === "metadata" && track.label === "thumbnails") thumbnailTrack = track;
				textTrackList.push({
					kind: track.kind,
					label: track.label,
					language: track.language,
					mode: track.mode
				});
				if ((track.kind === "captions" || track.kind === "subtitles") && track.mode === "showing") subtitlesShowing = true;
			}
			const chaptersCues = chaptersTrack?.cues ? Array.from(chaptersTrack.cues) : [];
			const thumbnailCues = thumbnailTrack?.cues ? Array.from(thumbnailTrack.cues) : [];
			let thumbnailTrackSrc = null;
			if (thumbnailTrack) thumbnailTrackSrc = findTrackElement(media, thumbnailTrack)?.src ?? null;
			for (const trackEl of media.querySelectorAll?.("track") ?? []) if (!trackEl.track?.cues?.length) listen(trackEl, "load", sync, { signal: trackCleanup.signal });
			set({
				chaptersCues,
				thumbnailCues,
				thumbnailTrackSrc,
				textTrackList,
				subtitlesShowing
			});
		}
		sync();
		listen(media.textTracks, "addtrack", sync, { signal });
		listen(media.textTracks, "removetrack", sync, { signal });
		listen(media.textTracks, "change", sync, { signal });
		listen(media, "loadstart", sync, { signal });
		signal.addEventListener("abort", () => trackCleanup?.abort(), { once: true });
	}
});

//#endregion
export { textTrackFeature };
