// @videojs/react/dist/default/ui/tooltip/tooltip-root.js
'use client';

import { useLatestRef } from "../../utils/use-latest-ref.js";
import { useSafeId } from "../../utils/use-safe-id.js";
import { TooltipContextProvider } from "./context.js";
import { useTooltipGroup } from "./group-context.js";
import { createTooltip, createTransition } from "@videojs/core/dom";
import { useSnapshot } from "@videojs/store/react";
import { useEffect, useState } from "react";
import { jsx } from "react/jsx-runtime";
import { TooltipCore, TooltipDataAttrs } from "@videojs/core";

//#region src/ui/tooltip/tooltip-root.tsx
function TooltipRoot({ open: controlledOpen, defaultOpen = TooltipCore.defaultProps.defaultOpen, onOpenChange: onOpenChangeProp, onOpenChangeComplete: onOpenChangeCompleteProp, delay = TooltipCore.defaultProps.delay, closeDelay = TooltipCore.defaultProps.closeDelay, disableHoverablePopup = TooltipCore.defaultProps.disableHoverablePopup, disabled = TooltipCore.defaultProps.disabled, children, ...coreProps }) {
	const [core] = useState(() => new TooltipCore(coreProps));
	core.setProps(coreProps);
	const isControlled = controlledOpen !== void 0;
	const groupFromContext = useTooltipGroup();
	const onOpenChangeRef = useLatestRef(onOpenChangeProp);
	const onOpenChangeCompleteRef = useLatestRef(onOpenChangeCompleteProp);
	const delayRef = useLatestRef(delay);
	const closeDelayRef = useLatestRef(closeDelay);
	const disableHoverablePopupRef = useLatestRef(disableHoverablePopup);
	const disabledRef = useLatestRef(disabled);
	const groupRef = useLatestRef(groupFromContext);
	const [tooltip] = useState(() => {
		const instance = createTooltip({
			transition: createTransition(),
			onOpenChange: (nextOpen, details) => {
				onOpenChangeRef.current?.(nextOpen, details);
			},
			onOpenChangeComplete: (nextOpen) => {
				onOpenChangeCompleteRef.current?.(nextOpen);
			},
			delay: () => delayRef.current,
			closeDelay: () => closeDelayRef.current,
			disableHoverablePopup: () => disableHoverablePopupRef.current,
			disabled: () => disabledRef.current,
			group: () => groupRef.current
		});
		if (!isControlled && defaultOpen) instance.open();
		return instance;
	});
	const anchorName = useSafeId();
	const popupId = useSafeId("tooltip-");
	useEffect(() => {
		if (controlledOpen === void 0) return;
		const { active: inputOpen } = tooltip.input.current;
		if (controlledOpen === inputOpen) return;
		if (controlledOpen) tooltip.open();
		else tooltip.close();
	}, [controlledOpen, tooltip]);
	useEffect(() => () => tooltip.destroy(), [tooltip]);
	const input = useSnapshot(tooltip.input);
	core.setInput(input);
	return /* @__PURE__ */ jsx(TooltipContextProvider, {
		value: {
			core,
			tooltip,
			state: core.getState(),
			stateAttrMap: TooltipDataAttrs,
			anchorName,
			popupId
		},
		children
	});
}

//#endregion
export { TooltipRoot };
