// @videojs/react/dist/dev/skins/dist/default/minimal/tailwind/video.tailwind.js
import { iconState } from "../../shared/tailwind/icon-state.js";
import { controls as controls$1 } from "./components/controls.js";
import { popup as popup$1 } from "./components/popup.js";
import { root as root$1 } from "./components/root.js";
import { bufferingIndicator } from "./components/buffering.js";
import { button } from "./components/button.js";
import { buttonGroup } from "./components/button-group.js";
import { error } from "./components/error.js";
import { icon, iconContainer, iconFlipped } from "./components/icon.js";
import { seek } from "./components/seek.js";
import { slider as slider$1 } from "./components/slider.js";
import { time } from "./components/time.js";
import { preview as preview$1 } from "./components/preview.js";
import { overlay } from "./components/overlay.js";
import { poster } from "./components/poster.js";
import { cn } from "@videojs/utils/style";

//#region ../skins/dist/default/minimal/tailwind/video.tailwind.js
const root = (isShadowDOM) => cn(root$1, "bg-black overflow-clip", "after:absolute after:pointer-events-none after:rounded-[inherit] after:z-10", "after:inset-0 after:ring-1 after:ring-inset after:ring-black/15 dark:after:ring-white/15", {
	"[&_::slotted(video)]:block [&_::slotted(video)]:w-full [&_::slotted(video)]:h-full [&_::slotted(video)]:rounded-(--media-video-border-radius) [&_::slotted(video)]:[object-fit:var(--media-object-fit,cover)] [&_::slotted(video)]:[object-position:var(--media-object-position,center)]": isShadowDOM,
	"[&_video]:block [&_video]:w-full [&_video]:h-full [&_video]:rounded-[inherit] [&_video]:[object-fit:var(--media-object-fit,contain)] [&_video]:[object-position:var(--media-object-position,center)]": !isShadowDOM
}, "[--media-video-border-radius:var(--media-border-radius,0.75rem)]", "[--media-controls-transition-duration:100ms]", "[--media-controls-transition-delay:0ms]", "[@media(pointer:fine)]:has-[[data-controls]:not([data-visible])]:[--media-controls-transition-delay:500ms]", "[@media(pointer:fine)]:has-[[data-controls]:not([data-visible])]:[--media-controls-transition-duration:300ms]", "[@media(pointer:coarse)]:has-[[data-controls]:not([data-visible])]:[--media-controls-transition-duration:150ms]", "motion-reduce:has-[[data-controls]:not([data-visible])]:[--media-controls-transition-duration:100ms]", "[--media-caption-track-y:-0.5rem]", "[--media-caption-track-delay:calc(var(--media-controls-transition-delay)_+_25ms)]", "[--media-caption-track-duration:var(--media-controls-transition-duration)]", "has-[[data-controls][data-visible]]:[--media-caption-track-y:-3rem]", !isShadowDOM ? [
	"[&_video::-webkit-media-text-track-container]:transition-[translate]",
	"[&_video::-webkit-media-text-track-container]:duration-(--media-caption-track-duration)",
	"[&_video::-webkit-media-text-track-container]:ease-out",
	"[&_video::-webkit-media-text-track-container]:delay-(--media-caption-track-delay)",
	"[&_video::-webkit-media-text-track-container]:translate-y-(--media-caption-track-y)",
	"[&_video::-webkit-media-text-track-container]:scale-98",
	"[&_video::-webkit-media-text-track-container]:z-1",
	"[&_video::-webkit-media-text-track-container]:font-[inherit]"
] : [], "[&:fullscreen]:[--media-border-radius:0]", {
	"[&:fullscreen_video]:object-contain": !isShadowDOM,
	"[&:fullscreen_::slotted(video)]:object-contain": isShadowDOM
});
const controls = cn(controls$1, "absolute bottom-0 inset-x-0", "pt-8 px-1.5 pb-1.5 gap-2", "[color:var(--media-color-primary,oklch(1_0_0))] z-10", "ease-out", "duration-(--media-controls-transition-duration)", "delay-(--media-controls-transition-delay)", "[@media(pointer:fine)]:will-change-[translate,filter,opacity]", "[@media(pointer:fine)]:transition-[translate,filter,opacity]", "[@media(pointer:coarse)]:will-change-[translate,opacity]", "[@media(pointer:coarse)]:transition-[translate,opacity]", "not-data-visible:opacity-0 not-data-visible:pointer-events-none", "motion-safe:not-data-visible:translate-y-full", "[@media(pointer:fine)]:motion-safe:not-data-visible:blur-sm", "@sm/media-root:pt-10 @sm/media-root:px-3 @sm/media-root:pb-3", "@sm/media-root:gap-3.5");
const preview = {
	...preview$1,
	root: cn("absolute left-(--media-slider-pointer) bottom-[calc(100%+0.5rem)] -translate-x-1/2", "opacity-0 scale-80 blur-sm origin-bottom", "transition-[scale,opacity,filter] duration-150", "group-data-pointing/slider:opacity-100 group-data-pointing/slider:scale-100 group-data-pointing/slider:blur-none", "[&:has([role=img][data-hidden])]:opacity-0 [&:has([role=img][data-hidden])]:scale-80 [&:has([role=img][data-hidden])]:blur-sm", "[&:has([role=img][data-loading])]:max-h-24", preview$1.root),
	thumbnailWrapper: cn(preview$1.thumbnailWrapper, "after:absolute after:inset-0 after:rounded-[inherit]", "after:ring-1 after:ring-black/5 after:shadow-sm after:shadow-black/20"),
	thumbnail: cn(preview$1.thumbnail, "max-w-44")
};
const slider = {
	...slider$1,
	track: cn(slider$1.track, "ring-1 ring-black/5")
};
const popup = {
	...popup$1,
	volume: cn(popup$1.popover, "[--media-popover-side-offset:0.5rem] p-1 bg-transparent")
};

//#endregion
export { controls, popup, preview, root, slider };
